Add a SQLite Interface for Web Extension storage

kiaraarose opened the pull request:
#### eec0bdebc484b8f58438640acf831ad486234388
<pre>
Add a SQLite Interface for Web Extension storage
<a href="https://bugs.webkit.org/show_bug.cgi?id=265830">https://bugs.webkit.org/show_bug.cgi?id=265830</a>

Reviewed by Timothy Hatcher.

This patch implements an Objective-C SQLite interface to be used for managing storage for web extensions.
This interface will be used for APIs such as Storage, Scripting, and Declarative Net Request.
_WKWebExtensionSQLiteStore will contain the generic methods needed for managing this storage,
and subclasses should be made for more requests for those APIs.

Testing:
Tested adding and removing data from storage for local test web extension.

* Source/WebKit/Configurations/WebKit.xcconfig:
Add libsqlite3.dylib to the WebKit framework.

* Source/WebKit/Platform/cocoa/CocoaHelpers.h:
* Source/WebKit/Platform/cocoa/CocoaHelpers.mm:
(WebKit::ensureDirectoryExists):
(WebKit::removeFileAtURL):
(WebKit::removeDirectoryIfEmpty):
(WebKit::separateResourceForkPath):
* Source/WebKit/Shared/Extensions/SQLiteDatatypeTraits.h: Added.
(WebKit::SQLiteDatatypeTraits&lt;int&gt;::fetch):
(WebKit::SQLiteDatatypeTraits&lt;int&gt;::bind):
(WebKit::SQLiteDatatypeTraits&lt;int64_t&gt;::fetch):
(WebKit::SQLiteDatatypeTraits&lt;int64_t&gt;::bind):
(WebKit::SQLiteDatatypeTraits&lt;double&gt;::fetch):
(WebKit::SQLiteDatatypeTraits&lt;double&gt;::bind):
(WebKit::SQLiteDatatypeTraits&lt;NSString::fetch):
(WebKit::SQLiteDatatypeTraits&lt;NSString::bind):
(WebKit::SQLiteDatatypeTraits&lt;NSData::fetch):
(WebKit::SQLiteDatatypeTraits&lt;NSData::bind):
(WebKit::SQLiteDatatypeTraits&lt;NSDate::fetch):
(WebKit::SQLiteDatatypeTraits&lt;NSDate::bind):
(WebKit::SQLiteDatatypeTraits&lt;NSNumber::fetch):
(WebKit::SQLiteDatatypeTraits&lt;NSNumber::bind):
(WebKit::SQLiteDatatypeTraits&lt;NSObject::fetch):
(WebKit::SQLiteDatatypeTraits&lt;NSObject::bind):
(WebKit::SQLiteDatatypeTraits&lt;std::nullptr_t&gt;::fetch):
(WebKit::SQLiteDatatypeTraits&lt;std::nullptr_t&gt;::bind):
(WebKit::SQLiteDatatypeTraits&lt;decltype):
* Source/WebKit/Shared/Extensions/SQLiteHelpers.h: Added.
(WebKit::SQLiteIsExecutionError):
(WebKit::SQLiteStatementBindAllParameters):
(WebKit::SQLiteDatabaseExecute):
(WebKit::SQLiteDatabaseExecuteAndReturnError):
(WebKit::_SQLiteDatabaseExecuteAndReturnError):
(WebKit::SQLiteDatabaseFetch):
(WebKit::_SQLiteStatementBindParameter):
(WebKit::_SQLiteStatementBindAllParameters):
(WebKit::_SQLiteStatementBindTupleParameters):
(WebKit::SQLiteEnumerationBlock::SQLiteEnumerationBlock):
(WebKit::SQLiteEnumerationBlock::_callBlockWithAllColumns):
(WebKit::SQLiteEnumerationBlock::callBlockWithAllColumns):
(WebKit::_StatementCallBlockWithAllColumns):
(WebKit::_WBSStatementFetchColumnsInTuple):
(WebKit::WBSStatementFetchColumnsInTuple):
(WebKit::_SQLiteStatementBindOrStep):
(WebKit::_SQLiteStatementBind):
(WebKit::SQLiteDatabaseEnumerate):
(WebKit::SQLiteDatabaseEnumerateRows):
* Source/WebKit/Shared/Extensions/_WKWebExtensionSQLiteDatabase.h: Added.
* Source/WebKit/Shared/Extensions/_WKWebExtensionSQLiteDatabase.mm: Added.
(-[_WKWebExtensionSQLiteDatabase initWithURL:queue:]):
(-[_WKWebExtensionSQLiteDatabase dealloc]):
(-[_WKWebExtensionSQLiteDatabase close]):
(-[_WKWebExtensionSQLiteDatabase reportErrorWithCode:query:error:]):
(-[_WKWebExtensionSQLiteDatabase reportErrorWithCode:statement:error:]):
(-[_WKWebExtensionSQLiteDatabase enableWAL:]):
(-[_WKWebExtensionSQLiteDatabase openWithAccessType:error:]):
(-[_WKWebExtensionSQLiteDatabase openWithAccessType:vfs:error:]):
(-[_WKWebExtensionSQLiteDatabase openWithAccessType:protectionType:vfs:error:]):
(-[_WKWebExtensionSQLiteDatabase _openWithFlags:vfs:error:]):
(+[_WKWebExtensionSQLiteDatabase inMemoryDatabaseURL]):
(+[_WKWebExtensionSQLiteDatabase privateOnDiskDatabaseURL]):
(+[_WKWebExtensionSQLiteDatabase _errorWithSQLiteErrorCode:userInfo:]):
* Source/WebKit/Shared/Extensions/_WKWebExtensionSQLiteRow.h: Added.
(-[_WKWebExtensionSQLiteRow uncopiedDataAtIndex:]):
* Source/WebKit/Shared/Extensions/_WKWebExtensionSQLiteRow.mm: Added.
(-[_WKWebExtensionSQLiteRow init]):
(-[_WKWebExtensionSQLiteRow initWithStatement:]):
(-[_WKWebExtensionSQLiteRow initWithCurrentRowOfEnumerator:]):
(-[_WKWebExtensionSQLiteRow stringAtIndex:]):
(-[_WKWebExtensionSQLiteRow intAtIndex:]):
(-[_WKWebExtensionSQLiteRow int64AtIndex:]):
(-[_WKWebExtensionSQLiteRow doubleAtIndex:]):
(-[_WKWebExtensionSQLiteRow boolAtIndex:]):
(-[_WKWebExtensionSQLiteRow dataAtIndex:]):
(-[_WKWebExtensionSQLiteRow uncopiedDataAtIndex:]):
(-[_WKWebExtensionSQLiteRow uncopiedRawDataAtIndex:]):
(-[_WKWebExtensionSQLiteRow _isNullAtIndex:]):
(-[_WKWebExtensionSQLiteRow objectAtIndex:]):
(-[_WKWebExtensionSQLiteRowEnumerator initWithResultsOfStatement:]):
(-[_WKWebExtensionSQLiteRowEnumerator nextObject]):
* Source/WebKit/Shared/Extensions/_WKWebExtensionSQLiteStatement.h: Added.
* Source/WebKit/Shared/Extensions/_WKWebExtensionSQLiteStatement.mm: Added.
(-[_WKWebExtensionSQLiteStatement init]):
(-[_WKWebExtensionSQLiteStatement initWithDatabase:query:error:]):
(-[_WKWebExtensionSQLiteStatement initWithDatabase:query:]):
(-[_WKWebExtensionSQLiteStatement dealloc]):
(-[_WKWebExtensionSQLiteStatement execute]):
(-[_WKWebExtensionSQLiteStatement execute:]):
(-[_WKWebExtensionSQLiteStatement fetch]):
(-[_WKWebExtensionSQLiteStatement fetchWithEnumerationBlock:error:]):
(-[_WKWebExtensionSQLiteStatement reset]):
(-[_WKWebExtensionSQLiteStatement invalidate]):
(-[_WKWebExtensionSQLiteStatement bindString:atParameterIndex:]):
(-[_WKWebExtensionSQLiteStatement bindInt:atParameterIndex:]):
(-[_WKWebExtensionSQLiteStatement bindInt64:atParameterIndex:]):
(-[_WKWebExtensionSQLiteStatement bindDouble:atParameterIndex:]):
(-[_WKWebExtensionSQLiteStatement bindData:atParameterIndex:]):
(-[_WKWebExtensionSQLiteStatement bindNullAtParameterIndex:]):
(-[_WKWebExtensionSQLiteStatement columnNamesToIndexes]):
(-[_WKWebExtensionSQLiteStatement columnNames]):
(-[_WKWebExtensionSQLiteStatement _isValid]):
* Source/WebKit/Shared/Extensions/_WKWebExtensionSQLiteStore.h: Added.
* Source/WebKit/Shared/Extensions/_WKWebExtensionSQLiteStore.mm: Added.
(-[_WKWebExtensionSQLiteStore dealloc]):
(-[_WKWebExtensionSQLiteStore deleteStorageWithCompletionHandler:]):
(-[_WKWebExtensionSQLiteStore deleteDatabaseWithCompletionHandler:]):
(-[_WKWebExtensionSQLiteStore useInMemoryDatabase]):
(-[_WKWebExtensionSQLiteStore _vacuum]):
(-[_WKWebExtensionSQLiteStore _openDatabaseIfNecessaryReturningErrorMessage:]):
(-[_WKWebExtensionSQLiteStore _openDatabase:deleteDatabaseFileOnError:]):
(-[_WKWebExtensionSQLiteStore _isDatabaseOpen]):
(-[_WKWebExtensionSQLiteStore _deleteDatabaseFileAtURL:reopenDatabase:]):
(-[_WKWebExtensionSQLiteStore _deleteExtensionStorageFolderIfEmpty]):
(-[_WKWebExtensionSQLiteStore _deleteDatabaseIfEmpty]):
(-[_WKWebExtensionSQLiteStore _deleteDatabase]):
(-[_WKWebExtensionSQLiteStore _handleSchemaVersioningWithDeleteDatabaseFileOnError:]):
(-[_WKWebExtensionSQLiteStore _migrateToCurrentSchemaVersionIfNeeded]):
(-[_WKWebExtensionSQLiteStore _setDatabaseSchemaVersion:]):
(-[_WKWebExtensionSQLiteStore _currentDatabaseSchemaVersion]):
(-[_WKWebExtensionSQLiteStore _databaseURL]):
(-[_WKWebExtensionSQLiteStore _createFreshDatabaseSchema]):
(-[_WKWebExtensionSQLiteStore _resetDatabaseSchema]):
(-[_WKWebExtensionSQLiteStore _isDatabaseEmpty]):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: <a href="https://commits.webkit.org/271574@main">https://commits.webkit.org/271574@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3454112f87d25f5152aa0f4c2626c6005ff151b7

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/28819 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/7462 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/30192 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/31438 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/26293 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/9596 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/4822 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/26345 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/29089 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/6184 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/24770 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/5374 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/5523 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/32777 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/26380 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/26222 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/31762 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/5496 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/3667 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/29547 "Found 1 new API test failure: /TestWTF:WTF_ParkingLot.UnparkOneOneFast (failure)") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/7117 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/5968 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/3719 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/6014 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

xeenon commented on Source/WebKit/Platform/cocoa/CocoaHelpers.mm:
```diff
@@ -357,6 +361,62 @@ id parseJSON(API::Data& json, JSONOptionSet options, NSError **error)
+    if (![NSFileManager.defaultManager createDirectoryAtPath:directory.path withIntermediateDirectories:YES attributes:nil error:&error])
+        return nil;
+    return directory;
+}
```
We should use `FileSystem::makeAllDirectories` from `#import <wtf/FileSystem.h>` for this.

xeenon commented on Source/WebKit/Platform/cocoa/CocoaHelpers.mm:
```diff
@@ -357,6 +361,62 @@ id parseJSON(API::Data& json, JSONOptionSet options, NSError **error)
+#endif
+
+    return YES;
+}
```
Ditto. `FileSystem::deleteFile`, otherwise we will need 3 more reviewers here.

xeenon commented on Source/WebKit/Platform/cocoa/CocoaHelpers.mm:
```diff
@@ -357,6 +361,62 @@ id parseJSON(API::Data& json, JSONOptionSet options, NSError **error)
+    ASSERT(URL.isFileURL);
+
+    return !rmdir(URL.fileSystemRepresentation);
+}
```
Ditto. `FileSystem::deleteEmptyDirectory`

xeenon commented on Source/WebKit/Platform/cocoa/CocoaHelpers.mm:
```diff
@@ -357,6 +361,62 @@ id parseJSON(API::Data& json, JSONOptionSet options, NSError **error)
+        return [path.stringByDeletingLastPathComponent stringByAppendingPathComponent:[@"._" stringByAppendingString:lastPathComponent]];
+
+    return nil;
+}
```
Not needed once we use `FileSystem` methods.

xeenon commented on Source/WebKit/Shared/Extensions/_WKWebExtensionSQLiteDatabase.mm:
```diff
@@ -0,0 +1,243 @@
+    return SQLiteDatabaseEnumerate(self, error, @"PRAGMA journal_mode = WAL", std::tie(std::ignore));
+}
+
+
```
Remove extra line.

xeenon commented on Source/WebKit/Shared/Extensions/_WKWebExtensionSQLiteDatabase.mm:
```diff
@@ -0,0 +1,243 @@
+    return NO;
+}
+
+
```
Remove extra line.

xeenon commented on Source/WebKit/Shared/Extensions/_WKWebExtensionSQLiteHelpers.h:
```diff
@@ -0,0 +1,369 @@
+    return result;
+}
+
+}
```
`} // namespace WebKit`

xeenon commented on Source/WebKit/Shared/Extensions/_WKWebExtensionSQLiteStore.mm:
```diff
@@ -0,0 +1,356 @@
+
+- (void)deleteDatabaseWithCompletionHandler:(void (^)(NSString *errorMessage))completionHandler
+{
+    auto weakSelf = WeakObjCPtr<_WKWebExtensionSQLiteStore>(self);
```
Since this is ObjC++ and ARC, you should just be able to do `__weak auto *weakSelf = self;`. Otherwise `WeakObjCPtr { self }` would be the simple form for `WeakObjCPtr`.

kiaraarose replied:
`__weak auto *weakSelf = self;` didn't work so i did `auto weakSelf = WeakObjCPtr<_WKWebExtensionSQLiteStore> { self };`

xeenon replied:
`WeakObjCPtr { self }` should work too, the type will be inferred.

kiaraarose replied:
I got an error when I did it without the type

xeenon replied:
Oh interesting! The other smart pointers work like that, so maybe they have some extra template magic that makes it work.

xeenon commented on Source/WebKit/Shared/Extensions/_WKWebExtensionSQLiteDatabase.mm:
```diff
@@ -0,0 +1,243 @@
+
+#if PLATFORM(IOS_FAMILY)
+    switch (protectionType) {
+    case SQLiteDatabaseProtectionType:
```
I think this line is the build failure on iOS, it can just be removed, I think.

kiaraarose replied:
was supposed to be `SQLiteDatabaseProtectionTypeDefault`. will update

xeenon commented on Source/WebKit/Platform/cocoa/CocoaHelpers.mm:
```diff
@@ -30,7 +30,13 @@
 #import "WKNSData.h"
+#import <wtf/FileSystem.h>
+
+#if PLATFORM(MAC)
+#import <sys/mount.h>
+#endif
```
Is this still needed?

kiaraarose replied:
nope, i'll remove

xeenon commented on Source/WebKit/Platform/cocoa/CocoaHelpers.mm:
```diff
@@ -357,6 +363,17 @@ id parseJSON(API::Data& json, JSONOptionSet options, NSError **error)
+NSURL *ensureDirectoryExists(NSURL *directory)
+{
+    ASSERT(directory.isFileURL);
+    if (!FileSystem::makeAllDirectories(directory.absoluteString)) {
```
`absoluteString` will be incorrect, since it will contain `file:///` prefix.

`FileSystem::makeAllDirectories(directory.path)`

xeenon replied:
This could also go in `wtf/NSURLExtras.mm/h`. Just need to drop the Extensions specific logging, which I would drop here if it stays anyway since CocoaHelpers is not extension specific.

kiaraarose committed:
Add a SQLite Interface for Web Extension storage
https://bugs.webkit.org/show_bug.cgi?id=265830

Reviewed by Timothy Hatcher.

This patch implements an Objective-C SQLite interface to be used for managing storage for web extensions.
This interface will be used for APIs such as Storage, Scripting, and Declarative Net Request.
_WKWebExtensionSQLiteStore will contain the generic methods needed for managing this storage,
and subclasses should be made for more requests for those APIs.

Testing:
Tested adding and removing data from storage for local test web extension.

* Source/WebKit/Configurations/WebKit.xcconfig:
Add libsqlite3.dylib to the WebKit framework.

* Source/WebKit/Platform/cocoa/CocoaHelpers.h:
* Source/WebKit/Platform/cocoa/CocoaHelpers.mm:
(WebKit::ensureDirectoryExists):
(WebKit::removeFileAtURL):
(WebKit::removeDirectoryIfEmpty):
(WebKit::separateResourceForkPath):
* Source/WebKit/Shared/Extensions/SQLiteDatatypeTraits.h: Added.
(WebKit::SQLiteDatatypeTraits<int>::fetch):
(WebKit::SQLiteDatatypeTraits<int>::bind):
(WebKit::SQLiteDatatypeTraits<int64_t>::fetch):
(WebKit::SQLiteDatatypeTraits<int64_t>::bind):
(WebKit::SQLiteDatatypeTraits<double>::fetch):
(WebKit::SQLiteDatatypeTraits<double>::bind):
(WebKit::SQLiteDatatypeTraits<NSString::fetch):
(WebKit::SQLiteDatatypeTraits<NSString::bind):
(WebKit::SQLiteDatatypeTraits<NSData::fetch):
(WebKit::SQLiteDatatypeTraits<NSData::bind):
(WebKit::SQLiteDatatypeTraits<NSDate::fetch):
(WebKit::SQLiteDatatypeTraits<NSDate::bind):
(WebKit::SQLiteDatatypeTraits<NSNumber::fetch):
(WebKit::SQLiteDatatypeTraits<NSNumber::bind):
(WebKit::SQLiteDatatypeTraits<NSObject::fetch):
(WebKit::SQLiteDatatypeTraits<NSObject::bind):
(WebKit::SQLiteDatatypeTraits<std::nullptr_t>::fetch):
(WebKit::SQLiteDatatypeTraits<std::nullptr_t>::bind):
(WebKit::SQLiteDatatypeTraits<decltype):
* Source/WebKit/Shared/Extensions/SQLiteHelpers.h: Added.
(WebKit::SQLiteIsExecutionError):
(WebKit::SQLiteStatementBindAllParameters):
(WebKit::SQLiteDatabaseExecute):
(WebKit::SQLiteDatabaseExecuteAndReturnError):
(WebKit::_SQLiteDatabaseExecuteAndReturnError):
(WebKit::SQLiteDatabaseFetch):
(WebKit::_SQLiteStatementBindParameter):
(WebKit::_SQLiteStatementBindAllParameters):
(WebKit::_SQLiteStatementBindTupleParameters):
(WebKit::SQLiteEnumerationBlock::SQLiteEnumerationBlock):
(WebKit::SQLiteEnumerationBlock::_callBlockWithAllColumns):
(WebKit::SQLiteEnumerationBlock::callBlockWithAllColumns):
(WebKit::_StatementCallBlockWithAllColumns):
(WebKit::_WBSStatementFetchColumnsInTuple):
(WebKit::WBSStatementFetchColumnsInTuple):
(WebKit::_SQLiteStatementBindOrStep):
(WebKit::_SQLiteStatementBind):
(WebKit::SQLiteDatabaseEnumerate):
(WebKit::SQLiteDatabaseEnumerateRows):
* Source/WebKit/Shared/Extensions/_WKWebExtensionSQLiteDatabase.h: Added.
* Source/WebKit/Shared/Extensions/_WKWebExtensionSQLiteDatabase.mm: Added.
(-[_WKWebExtensionSQLiteDatabase initWithURL:queue:]):
(-[_WKWebExtensionSQLiteDatabase dealloc]):
(-[_WKWebExtensionSQLiteDatabase close]):
(-[_WKWebExtensionSQLiteDatabase reportErrorWithCode:query:error:]):
(-[_WKWebExtensionSQLiteDatabase reportErrorWithCode:statement:error:]):
(-[_WKWebExtensionSQLiteDatabase enableWAL:]):
(-[_WKWebExtensionSQLiteDatabase openWithAccessType:error:]):
(-[_WKWebExtensionSQLiteDatabase openWithAccessType:vfs:error:]):
(-[_WKWebExtensionSQLiteDatabase openWithAccessType:protectionType:vfs:error:]):
(-[_WKWebExtensionSQLiteDatabase _openWithFlags:vfs:error:]):
(+[_WKWebExtensionSQLiteDatabase inMemoryDatabaseURL]):
(+[_WKWebExtensionSQLiteDatabase privateOnDiskDatabaseURL]):
(+[_WKWebExtensionSQLiteDatabase _errorWithSQLiteErrorCode:userInfo:]):
* Source/WebKit/Shared/Extensions/_WKWebExtensionSQLiteRow.h: Added.
(-[_WKWebExtensionSQLiteRow uncopiedDataAtIndex:]):
* Source/WebKit/Shared/Extensions/_WKWebExtensionSQLiteRow.mm: Added.
(-[_WKWebExtensionSQLiteRow init]):
(-[_WKWebExtensionSQLiteRow initWithStatement:]):
(-[_WKWebExtensionSQLiteRow initWithCurrentRowOfEnumerator:]):
(-[_WKWebExtensionSQLiteRow stringAtIndex:]):
(-[_WKWebExtensionSQLiteRow intAtIndex:]):
(-[_WKWebExtensionSQLiteRow int64AtIndex:]):
(-[_WKWebExtensionSQLiteRow doubleAtIndex:]):
(-[_WKWebExtensionSQLiteRow boolAtIndex:]):
(-[_WKWebExtensionSQLiteRow dataAtIndex:]):
(-[_WKWebExtensionSQLiteRow uncopiedDataAtIndex:]):
(-[_WKWebExtensionSQLiteRow uncopiedRawDataAtIndex:]):
(-[_WKWebExtensionSQLiteRow _isNullAtIndex:]):
(-[_WKWebExtensionSQLiteRow objectAtIndex:]):
(-[_WKWebExtensionSQLiteRowEnumerator initWithResultsOfStatement:]):
(-[_WKWebExtensionSQLiteRowEnumerator nextObject]):
* Source/WebKit/Shared/Extensions/_WKWebExtensionSQLiteStatement.h: Added.
* Source/WebKit/Shared/Extensions/_WKWebExtensionSQLiteStatement.mm: Added.
(-[_WKWebExtensionSQLiteStatement init]):
(-[_WKWebExtensionSQLiteStatement initWithDatabase:query:error:]):
(-[_WKWebExtensionSQLiteStatement initWithDatabase:query:]):
(-[_WKWebExtensionSQLiteStatement dealloc]):
(-[_WKWebExtensionSQLiteStatement execute]):
(-[_WKWebExtensionSQLiteStatement execute:]):
(-[_WKWebExtensionSQLiteStatement fetch]):
(-[_WKWebExtensionSQLiteStatement fetchWithEnumerationBlock:error:]):
(-[_WKWebExtensionSQLiteStatement reset]):
(-[_WKWebExtensionSQLiteStatement invalidate]):
(-[_WKWebExtensionSQLiteStatement bindString:atParameterIndex:]):
(-[_WKWebExtensionSQLiteStatement bindInt:atParameterIndex:]):
(-[_WKWebExtensionSQLiteStatement bindInt64:atParameterIndex:]):
(-[_WKWebExtensionSQLiteStatement bindDouble:atParameterIndex:]):
(-[_WKWebExtensionSQLiteStatement bindData:atParameterIndex:]):
(-[_WKWebExtensionSQLiteStatement bindNullAtParameterIndex:]):
(-[_WKWebExtensionSQLiteStatement columnNamesToIndexes]):
(-[_WKWebExtensionSQLiteStatement columnNames]):
(-[_WKWebExtensionSQLiteStatement _isValid]):
* Source/WebKit/Shared/Extensions/_WKWebExtensionSQLiteStore.h: Added.
* Source/WebKit/Shared/Extensions/_WKWebExtensionSQLiteStore.mm: Added.
(-[_WKWebExtensionSQLiteStore dealloc]):
(-[_WKWebExtensionSQLiteStore deleteStorageWithCompletionHandler:]):
(-[_WKWebExtensionSQLiteStore deleteDatabaseWithCompletionHandler:]):
(-[_WKWebExtensionSQLiteStore useInMemoryDatabase]):
(-[_WKWebExtensionSQLiteStore _vacuum]):
(-[_WKWebExtensionSQLiteStore _openDatabaseIfNecessaryReturningErrorMessage:]):
(-[_WKWebExtensionSQLiteStore _openDatabase:deleteDatabaseFileOnError:]):
(-[_WKWebExtensionSQLiteStore _isDatabaseOpen]):
(-[_WKWebExtensionSQLiteStore _deleteDatabaseFileAtURL:reopenDatabase:]):
(-[_WKWebExtensionSQLiteStore _deleteExtensionStorageFolderIfEmpty]):
(-[_WKWebExtensionSQLiteStore _deleteDatabaseIfEmpty]):
(-[_WKWebExtensionSQLiteStore _deleteDatabase]):
(-[_WKWebExtensionSQLiteStore _handleSchemaVersioningWithDeleteDatabaseFileOnError:]):
(-[_WKWebExtensionSQLiteStore _migrateToCurrentSchemaVersionIfNeeded]):
(-[_WKWebExtensionSQLiteStore _setDatabaseSchemaVersion:]):
(-[_WKWebExtensionSQLiteStore _currentDatabaseSchemaVersion]):
(-[_WKWebExtensionSQLiteStore _databaseURL]):
(-[_WKWebExtensionSQLiteStore _createFreshDatabaseSchema]):
(-[_WKWebExtensionSQLiteStore _resetDatabaseSchema]):
(-[_WKWebExtensionSQLiteStore _isDatabaseEmpty]):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: https://commits.webkit.org/271574@main
